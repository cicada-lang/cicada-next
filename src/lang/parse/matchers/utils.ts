import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
