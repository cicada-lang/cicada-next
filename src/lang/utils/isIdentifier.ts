const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
