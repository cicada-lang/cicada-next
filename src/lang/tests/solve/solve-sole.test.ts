import { expect, test } from "vitest"
import { runCode } from "../utils"

test("solve Sole", async () => {
  const output = await runCode(`

solve () {
  unify sole = sole
  unify sole = sole
  unify sole = sole
}

`)

  expect(output).toMatchInlineSnapshot('"{  }"')
})

test("solve Sole", async () => {
  const output = await runCode(`

solve (a: Trivial, b: Trivial, c: Trivial) {
  unify a = b
  unify b = c
  unify c = sole
}

`)

  expect(output).toMatchInlineSnapshot('"{ a: sole, b: sole, c: sole }"')
})
