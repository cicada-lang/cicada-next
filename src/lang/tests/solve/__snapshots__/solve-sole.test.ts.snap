// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`solve Sole -- abc 1`] = `
"{
  a: sole,
  b: sole,
  c: sole
}"
`;

exports[`solve Sole 1`] = `"{}"`;
